File: atcoder/abc274_e_TSP_bitDP.py
# https://atcoder.jp/contests/abc274/tasks/abc274_e
# # def input(): return sys.stdin.readline().rstrip()
# # input = sys.stdin.readline
# from numba import njit
# from functools import lru_cache

import sys
input = sys.stdin.buffer.readline
# sys.setrecursionlimit(10 ** 7)
INF = 1e18

N, M = map(int, input().split())
W = N+M+1
pos = []
for i in range(N+M):
    pos.append(tuple(map(int, input().split())))
pos.append((0,0)) # 原点
# print(pos)

Wbit = 1<<W
dp = [[INF]*W for _ in range(Wbit)]
dp[0][N+M] = 0
sp = [1.0]*Wbit
for s in range(Wbit):
    for v in range(N, N+M):
        if ((s>>v)&1): sp[s] /= 2
# print(sp)

dist = [[INF]*W for _ in range(W)]
for v in range(W):
    for u in range(W):
        dist[v][u] = ((pos[v][0]-pos[u][0])**2 + (pos[v][1]-pos[u][1])**2)**0.5
# print(dist)

for s in range(Wbit):
    for v in range(W):
        if dp[s][v]==INF: continue
        for u in range(W):
            if ((s>>u)&1): continue
            dp[s|(1<<u)][u] = min(dp[s|(1<<u)][u], dp[s][v] + dist[v][u]*sp[s])
# print(dp)

t = (1<<N)-1 | (1<<(N+M))

ans = INF
for s in range(Wbit):
    if ((s&t)==t):
        ans = min(ans, dp[s][N+M])
print(ans)



# TLE
# import sys
# input = sys.stdin.buffer.readline
# sys.setrecursionlimit(10 ** 7)

# N, M = map(int, input().split())
# XY = [[int(xy) for xy in input().split()] for _ in range(N)]
# # print(XY)
# PQ = [[int(pq) for pq in input().split()] for _ in range(M)]
# # print(PQ)

# ans = 1001001001001.0

# def dfs(dist, x, y, cnt, speed, gone):
#     global ans
#     if cnt==N:
#         dist += ((x*x + y*y)**0.5)/speed
#         # print(ans, dist, cnt, speed, gone)
#         if dist<ans: ans = dist
#         return
#     for i in range(N+M):
#         if gone[i]: continue
#         gone[i] = True
#         if i<N:
#             nx, ny = XY[i]
#             move = (((nx-x)**2 + (ny-y)**2)**0.5)/speed
#             dfs(dist + move, nx, ny, cnt + 1, speed, gone)
#         else:
#             nx, ny = PQ[i-N]
#             move = (((nx-x)**2 + (ny-y)**2)**0.5)/speed
#             dfs(dist + move, nx, ny, cnt, speed * 2, gone)
#         gone[i] = False
#     return

# dfs(0, 0, 0, 0, 1, [False]*(N+M))
# print(ans)

File: atcoder/abc274_e_TSP_bitDP.cpp
// https://atcoder.jp/contests/abc274/tasks/abc274_e
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
// #include <atcoder/all>
// using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

// Geometry
const double eps = 1e-9;
bool equal(double a, double b) { return abs(a-b) < eps;}
 
// Vector
// https://youtu.be/UWbGRhF3Ozw?t=9564
struct V {
  double x, y;
  V(double x=0, double y=0): x(x), y(y) {}
  V& operator+=(const V& v) { x += v.x; y += v.y; return *this;}
  V operator+(const V& v) const { return V(*this) += v;}
  V& operator-=(const V& v) { x -= v.x; y -= v.y; return *this;}
  V operator-(const V& v) const { return V(*this) -= v;}
  V& operator*=(double s) { x *= s; y *= s; return *this;}
  V operator*(double s) const { return V(*this) *= s;}
  V& operator/=(double s) { x /= s; y /= s; return *this;}
  V operator/(double s) const { return V(*this) /= s;}
  double dot(const V& v) const { return x*v.x + y*v.y;}
  double cross(const V& v) const { return x*v.y - v.x*y;}
  double norm2() const { return x*x + y*y;}
  double norm() const { return sqrt(norm2());}
  V rotate90() const { return V(y, -x);}
  int ort() const { // orthant
    if (abs(x) < eps && abs(y) < eps) return 0;
    if (y > 0) return x>0 ? 1 : 2;
    else return x>0 ? 4 : 3;
  }
  bool operator<(const V& v) const {
    int o = ort(), vo = v.ort();
    if (o != vo) return o < vo;
    return cross(v) > 0;
  }
};
istream& operator>>(istream& is, V& v) {
  is >> v.x >> v.y; return is;
}
ostream& operator<<(ostream& os, const V& v) {
  os<<"("<<v.x<<","<<v.y<<")"; return os;
}

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
    int n, m;
    cin >> n >> m;
    int w = n+m+1;
    vector<V> p(w);
    rep(i,w-1) cin >> p[i];
    int w2 = 1<<w;
    vector<double> sp(w2,1);
    rep(s,w2) {
        rep(i,m) if(s>>(n+i)&1) sp[s] /= 2;
    }

    vector dist(w, vector<double>(w));
    rep(i,w) rep(j,w) dist[i][j] = (p[i]-p[j]).norm();

    const double INF = 1e18;
    vector dp(w2, vector<double>(w,INF));
    dp[0][n+m] = 0;
    rep(s,w2) rep(v,w) {
        if (dp[s][v]==INF) continue;
        rep(u,w) if (~s>>u&1) {
            chmin(dp[s|1<<u][u], dp[s][v]+dist[v][u]*sp[s]);
        }
    }

    int t = (1<<n)-1 | (1<<(n+m));
    double ans = INF;
    rep(s,w2) if ((s&t)==t) chmin(ans, dp[s][n+m]);
    printf("%.10f\n", ans);

	return 0;
}
